src/string_array.c

```diff
@@ -133,6 +134,68 @@ rcutils_string_array_cmp(
   return RCUTILS_RET_OK;
 }
 
+rcutils_ret_t
+rcutils_string_array_resize(
+  rcutils_string_array_t * string_array,
+  size_t new_size)
+{
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    string_array, "string_array is null", return RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (string_array->size == new_size) {
+    return RCUTILS_RET_OK;
+  }
+
+  rcutils_allocator_t * allocator = &string_array->allocator;
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(
+    allocator, "allocator is invalid", return RCUTILS_RET_INVALID_ARGUMENT);
+
+  // Stash entries being removed
+  rcutils_string_array_t to_reclaim = rcutils_get_zero_initialized_string_array();
+  if (new_size < string_array->size) {
+    size_t num_removed = string_array->size - new_size;
+    rcutils_ret_t ret = rcutils_string_array_init(&to_reclaim, num_removed, allocator);
+    if (RCUTILS_RET_OK != ret) {
+      // rcutils_string_array_init should have already set an error message
+      return ret;
+    }
+    for (size_t i = 0; i < to_reclaim.size; ++i) {
+      to_reclaim.data[i] = string_array->data[new_size + i];
+    }
```

Comment:
nitpick only, but I think that could be a single `memcpy` line:
```suggestion
    std::memcpy(to_reclaim.data, string_array->data[new_size], to_reclaim.size);
```